### PR TITLE
[TwoAddressInstruction] Update LiveIntervals after INSERT_SUBREG with undef read

### PR DESCRIPTION
**Pull request by jayfoad**
Update LiveIntervals after rewriting:
  %reg = INSERT_SUBREG undef %reg, %subreg, subidx
to:
  undef %reg:subidx = COPY %subreg

D113044 implemented this for the non-undef case.

### CODE REVIEW

File: llvm/lib/CodeGen/TwoAddressInstructionPass.cpp [hunk 1]
```diff
@@ -1868,12 +1868,16 @@ bool TwoAddressInstructionPass::runOnMachineFunction(MachineFunction &Func) {
...
-            SlotIndex Idx = LIS->getInstructionIndex(*mi);
+            SlotIndex Idx = LIS->getInstructionIndex(*mi).getRegSlot();
             for (auto &S : LI.subranges()) {
               if ((S.LaneMask & LaneMask).none()) {
-                LiveRange::iterator UseSeg = S.FindSegmentContaining(Idx);
-                LiveRange::iterator DefSeg = std::next(UseSeg);
-                S.MergeValueNumberInto(DefSeg->valno, UseSeg->valno);
+                LiveRange::iterator DefSeg = S.FindSegmentContaining(Idx);
+                if (mi->getOperand(0).isUndef()) {
+                  S.removeValNo(DefSeg->valno);
```
Comment (jayfoad) [inline llvm/lib/CodeGen/TwoAddressInstructionPass.cpp:1876]: TBH I am pretty nervous about all this low level fiddling around with live ranges. I would be much happier if there was some higher level API I could use to do this kind of update.

**Comment by jayfoad** [comment]
I see #65916 has just been merged which fixes exactly the same kind of problem.

**Comment by lukel97** [comment]
> I see #65916 has just been merged which fixes exactly the same kind of problem.

It was just closed, not merged. I cherry picked this onto #65997 and can confirm it fixes the crash, thanks!

**Review by qcolombet — APPROVED** [review]
LGTM

**Pull request merged**